iotdb/src/main/java/org/apache/iotdb/db/query/executor/GroupByWithValueFilterDataSet.java

```diff
@@ -0,0 +1,145 @@
+import org.apache.iotdb.tsfile.utils.Pair;
+
+public class GroupByWithValueFilterDataSet extends GroupByEngine {
+
+
+  private List<EngineReaderByTimeStamp> allDataReaderList;
+  private TimeGenerator timestampGenerator;
+  private long timestamp;
+  private boolean hasCachedTimestamp;
+
+  //group by batch calculation size.
+  private int timeStampFetchSize;
+
+  /**
+   * constructor.
+   */
+  public GroupByWithValueFilterDataSet(long jobId, List<Path> paths, long unit, long origin,
+      List<Pair<Long, Long>> mergedIntervals) {
+    super(jobId, paths, unit, origin, mergedIntervals);
+    this.allDataReaderList = new ArrayList<>();
+    this.timeStampFetchSize = 10 * IoTDBDescriptor.getInstance().getConfig().getFetchSize();
+  }
+
+  /**
+   * init reader and aggregate function.
+   */
+  public void initGroupBy(QueryContext context, List<String> aggres, IExpression expression)
+      throws FileNodeManagerException, PathErrorException, ProcessorException, IOException {
+    initAggreFuction(aggres);
+
+    QueryTokenManager.getInstance().beginQueryOfGivenExpression(jobId, expression);
+    QueryTokenManager.getInstance().beginQueryOfGivenQueryPaths(jobId, selectedSeries);
+    this.timestampGenerator = new EngineTimeGenerator(jobId, expression, context);
+    this.allDataReaderList = SeriesReaderFactory
+        .getByTimestampReadersOfSelectedPaths(jobId, selectedSeries, context);
+  }
+
+  @Override
+  public RowRecord next() throws IOException {
+    if (!hasCachedTimeInterval) {
+      throw new IOException(
+          "need to call hasNext() before calling next() in GroupByWithOnlyTimeFilterDataSet.");
+    }
+    hasCachedTimeInterval = false;
+    for (AggregateFunction function : functions) {
+      function.init();
+    }
+
+    long[] timestampArray = new long[timeStampFetchSize];
+    int timeArrayLength = 0;
+    if (hasCachedTimestamp) {
+      if (timestamp < endTime) {
+        hasCachedTimestamp = false;
+        timestampArray[timeArrayLength++] = timestamp;
+      } else {
+        //所有域均为空
+        return constructRowRecord();
+      }
+    }
+
+    while (timestampGenerator.hasNext()) {
+      //construct timestamp list
+      for (int cnt = 1; cnt < timeStampFetchSize; cnt++) {
+        if (!timestampGenerator.hasNext()) {
+          break;
+        }
+        timestamp = timestampGenerator.next();
+        if (timestamp < endTime) {
+          timestampArray[timeArrayLength++] = timestamp;
+        } else {
+          hasCachedTimestamp = true;
+          break;
+        }
+      }
+
+      //cal result using timestamp list
+      for (int i = 0; i < selectedSeries.size(); i++) {
+        functions.get(i).calcAggregationUsingTimestamps(
+            timestampArray, timeArrayLength, allDataReaderList.get(i));
+      }
+
+      timeArrayLength = 0;
+      //judge if it's end
+      if (timestamp >= endTime) {
+        hasCachedTimestamp = true;
+        break;
+      }
+    }
+
+    if (timeArrayLength > 0) {
```

Comment:
Is these judges needed?
I think timeArrayLength always equals 0?

Reply:
when hasCachedTimestamp==true and !timestampGenerator.hasNext(), if doesn't have this judge, the cached timestamp won't be calculated.